### DISCUSSION TITLE
Advanced Search Question

### DISCUSSION THREAD

**Discussion by jcurry5** [nextgenhealthcare/connect#5334] [General]
I'm trying to search on the Sent message type for: Value=""

But when I do so I get no returns.  Searching for Value= returns everything so I'm assuming there needs to be an escape character for quotes?  If so, what should be used?  Thanks.

**Comment by mjtrac** [comment]
I'm not sure I understand your question, but entering "" into the text search area in basic search works fine, and in the advanced search area, creating a search on, for example, Raw message, with just "" works fine. Version 4.0.0 
![image](https://user-images.githubusercontent.com/2815700/183132194-d0d5f392-eca7-46c3-aec9-b8c726d451eb.png)
![image](https://user-images.githubusercontent.com/2815700/183132345-56f6173e-a134-4711-be1a-1d537d3aeb01.png)

**Reply by jcurry5** [reply]
My Content Type is 'Sent' so when I search for "" it returns no results.  My 'Sent' has a string of Value="" however searching for Value="" returns nothing when I expect  to get results.  Even if I search for Value=" it still returns no messages when it should.  So I'm guessing that quotes needs to be terminated in some fashion.

**Reply by jonbartels** [reply]
https://regex101.com/r/xIr4Lv/1

Try a regex search using `Value=\"\"`

**Reply by mjtrac** [reply]
And if the regex box is disabled, it probably means you're still using Derby as your database, in which case Regex search is not available AFAIK.  (I just tried a test without enabling Regex and \\" didn't match ".  I was surprised that " didn't match either, when searching in the Advanced Search on type Sent, at least.)  In any event, you don't want to use Derby in production.

**Reply by jcurry5** [reply]
I'm not using Derby.  It's against a SQL server in production.  Simply putting Value="" just doesn't work.  I even tried regex in the Advanced Search field for 'Sent' using Value=\"\" and no values are returned either.  I'm kind of scratching my head as to why this isn't working.  I can't profiler the query right now to see how SQL is encapsulating the quotes but I suspect that is the culprit.

**Reply by jcurry5** [reply]
So I dug a little deeper in how the CONTENT is stored and it's stored like this in the Mirth database:

```
  <entry>
    <string>value</string>
    <string></string>
  </entry>
```

My thinking is that when the data is stored like this it'll never return true.  Even though the Sent is setup as Value="".

**Reply by mjtrac** [reply]
I'd never say never, but if it shows with an upper case V in your "View Messages" listing, when you are looking at the destination's Sent value, then it should match an upper case V.  If you are using a different data type, you should still be able to look at this view and see what got stored, I'd think.

![image](https://user-images.githubusercontent.com/2815700/183141130-a53a63ef-8bf3-4f02-a948-c2998ba4354d.png)

**Reply by jcurry5** [reply]
Sorry that was me typing it out.  It is the right case but I did try Value=\"\", value=\"\", Value="", and Value=\"\" but still nothing.

**Reply by jcurry5** [reply]
Sorry I take that back.  I double checked and what's in the Sent message is Value="" but what's in SQL Content is

```
<entry>
    <string>value</string>
    <string></string>
  </entry>
```
I'd like to think that wouldn't be case sensitive though.  Even looking at SQL profiler it's based off of what you're putting in the search content box.

Here's how SQL is building the query:

```
declare @p1 int
set @p1=26
exec sp_prepare @p1 output,N'@P0 bigint,@P1 bigint,@P2 int,@P3 nvarchar(4000)',N'SELECT MESSAGE_ID AS MESSAGE_ID, METADATA_ID AS METADATA_ID, CONTENT_TYPE AS CONTENT_TYPE
			
		FROM D_MC26
		
		 WHERE MESSAGE_ID  >=   @P0 
			
			
				AND MESSAGE_ID  <=   @P1 
			
			
			
				AND CONTENT_TYPE =  @P2 
			
			
				AND CONTENT LIKE ''%'' +  @P3  + ''%''',1
select @p1

exec sp_execute 26,1,50,1,N'Value=""'
```

**Reply by jcurry5** [reply]
Seeing as how the data is stored in XML format, any search for Value="" will always return false.  Because of the literal string of Value="" doesn't exist.  'Value' is a variable in my Mirth destination so it creates a new <string> as part of the XML builder.  And then a value is either assigned or not:

`<entry>
    <string>value</string>
    <string></string>
</entry>`

Because Value="" isn't in any part of the CONTENT column it'll never return true.  I'm just wondering if I should try to search for <string>value</string><string></string>...

**Reply by jcurry5** [reply]
The rabbit hole goes deeper...

When I do an advanced search for 'Sent' against:
`<string>value</string>%<string></string>`

I get no results

When I do the same search from the standard Text Box, I get results, albeit everything.  What does this tell me?  I have no clue how Mirth does its searching and why there are so many discrepancies between Text box search, Advanced Search, how it stores its data and how it builds the searches through SQL.

**Reply by jonbartels** [reply]
![image](https://user-images.githubusercontent.com/7669292/183153717-27f3119d-acd2-4049-b562-85a008ffa76f.png)

Does searching for content contains help?

**Reply by jcurry5** [reply]
Nope - I've tried that and it doesn't do anything.  From what I'm understanding (which could be wrong), I have a variable of "Value" that gets assigned in the destination transformer.  As part of that transformer, the Source has an XML template that reads "Value=${value}".  So what's getting stored in Mirth is <string>value</string><string>*value of string*</string> and that's how it's stored in SQL.  I don't know if Mirth tries to break apart the XML content and rebuild but I'm assuming it's not and strictly doing a search for whatever is in Contains.  Which isn't even remotely right because of how the data is stored from the Source template.

**Reply by jcurry5** [reply]
I guess it took going back and forth but I finally figured it out.

I retract my previous comment.

So it looks like it searches different content types.  I was strictly looking at content type 10 but when I searched on content type 5 (essentially what you see in the Sent radio), I was able to find what I'm looking for.

`Value=&quot;&quot`

This will now return anything of Value=""

Thanks - hopes this helps anyone from the future.

**Reply by jonbartels** [reply]
The XML content shown is how Mirth serializes its map objects (channelMap, connectorMap, etc). 

Something like this *might* work but you'd have to account for the content being pretty printed in the DB or not:
![image](https://user-images.githubusercontent.com/7669292/183156838-93d41ae6-dd4e-47e2-86c1-c6b8105280f7.png)

If this is a one-time task you might be better off querying the DB directly.

**Reply by jcurry5** [reply]
Yeah I tried the XML serialized approach but still no go.  Even throwing in char(10), char(13), wildcards, it still wasn't working.  What did work was my previous comment by putting in:
 `Value=&quot;&quot`
